### ISSUE TITLE
ITMS-90338: Non-public API usage - The app references non-public selectors in [App]

### ISSUE THREAD

**Issue by rutledmj** [xamarin/Xamarin.Forms#15128]
Hi, I am building a Xamarin.Forms app that is building properly and deploying to the Apple App Store but keeps getting rejected for the following reasons:

> ITMS-90338: Non-public API usage - The app references non-public selectors in [App].Mobile.iOS: accessibilityBrailleMapRenderRegion, accessibilityBrailleMapRenderer, accessibilityColumnRange, accessibilityRowRange, activePlayer, beginAuthorizationWithRequest:, cancelAuthorizationWithRequest:, errorForCHIPErrorCode:, errorToCHIPErrorCode:, playerId, serviceExtensionWillTerminate, session:didReceiveData:fromPlayer:, session:didReceiveMessage:withData:fromPlayer:, session:player:didChangeConnectionState:, session:player:didSaveData:, setAccessibilityBrailleMapRenderRegion:, setAccessibilityBrailleMapRenderer:. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app.

I coded the app on a PC in Visual Studio Community 2019 v16.9.1.  And built and deployed the app via Appcenter.ms. 

Packages that I am using are: 
Microsoft.AppCenter.Analytics v4.5.0
Microsoft.AppCenter.Crashes v4.5.0
Xamarin.Essentials v1.7.1
Xamarin.Forms v5.0.0.2337

Target Framework: .NET Standard 2.0

Information that I have confirmed from other sources:
Build Configuration: Release
Linker Behavior: Link Framework SDKs Only
Additional mtouch arguments: --linkskip=[selectors], --linkplatform

I have found all of the non-public selectors in xamarin/xamarin-macios and xamarin/apple-api-docs but I have not been able to find out how to fix the errors.

Thanks

**Comment by jfversluis** [comment]
Hey there! Why are you adding `--linkskip=[selectors]` this? I think that is your problem :)

**Comment by rutledmj** [comment]
Thanks for your advice. I ran into the issue prior to adding that to the mtouch arguments. The issue occurred both with and without the arguments. 

I ended up getting a Mac VM and publishing the app through Transporter. 
The issue is definitely with the Xamarin.iOS SDK. When I change versions, the error was the same, but the arguments changed.

**Comment by jfversluis** [comment]
Thanks for letting us know! Raised it with the iOS team, closing it here as there is nothing we can do on our side :)

**Comment by evgenyKozlov1996** [comment]
Hello, @rutledmj 

I faced with the same issue, with the same list of private APIs but I am not use any AppCenter SDKs. I have the following links:
- Newtonsoft.Json 12.0.3 
- ReactiveUI.XamForms 14.1.1
- ReactiveUI 14.1.1
- RestSharp 106.11.7
- Splat.Microsoft.Extensions.DependencyInjection 12.1.4
- Xamarin.CommunityToolkit 1.1.0
- Xamarin.Essentials.Interfaces 1.7.0
- Xamarin.Forms.Visual.Material 5.0.0.2125
- Xamarin.Forms 5.0.0.2125
- Acr.UserDialogs 7.2.0.534
- Microsoft.EntityFrameworkCore.Sqlite 3.1.15
- Microsotf.EntityFrameworkCore.proxies 3.1.15.

But I am confused a lot by accessibilityBrailleMapRenderer field. Looks like it's a very specific thing that should be a part of Xamarin.IOS...

I tried to remove CommunityToolkit, Visual.Material and UserDialogs but the issue still happened. But, if you create a new Xamarin Forms application, without any libraries, it will be published successfully. Maybe there are some libraries that are causing this error...

**Comment by rutledmj** [comment]
I built the app on Visual Studio on PC and tried to publish through Appcenter. These options aren't available on PC as far as I know.

**Comment by melissa-gzz** [comment]
yes in vs pc, when you build the app you need to select on top the option release instead of debug, and in the next select the option that says iphone, at least that work for me

**Comment by rutledmj** [comment]
Oh I'm sorry. VS for PC doesn't have archiving as far as I know.

I did set the build configuration to Release and Device to iPhone.

**Comment by slongietti** [comment]
I also believe this to be an Xamarin.iOS.SDK issue. 

I pulled code from our repo at a commit that previously worked. Without changing any code or csproj settings, I ran it through our unchanged pipeline on Azure DevOps that published the application successfully prior to 2/4/2022. The result was the error stated in this issue.

I then attempted to build and publish locally on Mac using VS 2019 for Mac. I first created the ipa file locally. Then I pushed the application via Transporter. Still the results were the same.

@xamarin-release-manager Can we please re-open this issue or create an issue that reflects what is going on?

Please let me know if I can provide more information. 

Xamarin versions attempted: 

Xamarin-Forms: 5.0.0.2291 & 5.0.0.2337
Xamarin-Forms-Maps: 5.0.0.2291 & 5.0.0.2337

**Comment by slongietti** [comment]
UPDATE: This is now working with the new 15.6.0.3 Xamarin.iOS SDK or whatever iOS did in the 15.3.1 update that "coincidentally" lists one change for Braille

This seems to convenient not to be related when the error message related to this issue is littered with methods for braille. (bold below).

ITMS-90338: Non-public API usage - The app references non-public selectors in [App].Mobile.iOS: **accessibilityBrailleMapRenderRegion**, **accessibilityBrailleMapRenderer**, accessibilityColumnRange, accessibilityRowRange, activePlayer, beginAuthorizationWithRequest:, cancelAuthorizationWithRequest:, errorForCHIPErrorCode:, errorToCHIPErrorCode:, playerId, serviceExtensionWillTerminate, session:didReceiveData:fromPlayer:, session:didReceiveMessage:withData:fromPlayer:, session:player:didChangeConnectionState:, session:player:didSaveData:, **setAccessibilityBrailleMapRenderRegion**:, **setAccessibilityBrailleMapRenderer**:. If method names in your source code match the private Apple APIs listed above, altering your method names will help prevent this app from being flagged in future submissions. In addition, note that one or more of the above APIs may be located in a static library that was included with your app.

**Comment by slongietti** [comment]
I was able to resolve this by changing the Linker setting from None to SdkOnly. Although the build I mentioned above did work with the Linker setting set to None, it no longer does and SdkOnly is recommended by Xamarin. An extra bonus is SdkOnly will reduce the size of the app.

However when I changed the Linker setting to SdkOnly, I received a new error described here:

https://docs.microsoft.com/en-us/answers/questions/369135/xamarin-ios-app-healthkit-reject-and-custommarshal.html

Luckily I was able to resolve this by adding a custom linker file which is described in my answer in post above.

![image](https://user-images.githubusercontent.com/33959892/155557151-891dbb1a-37cb-47ab-99b0-4460f460d76e.png)
